### PR TITLE
Feature/unfulfilledpledges

### PR DESCRIPTION
**Pull request by harshvitra**
Tasks 
- [x] Design New card for unfulfilled Pledges
- [x] Design new buttons for payment and add more trees
- [x] Add Anonymous option in the form and reflect in the pledges
- [x] Get pre-filled data in the form for logged in users
- [x] Storing pledges/pledge Token of user in local storage 
- [x] Storing token in user profile for logged in user (API not ready)

### CODE REVIEW

**Comment by harshvitra** [comment]
@sagararyal you can use this for creating the video for client

**Comment by sagararyal** [comment]
- [ ] Clicking on Event, does not load the pledge and offer to increase it (it should not show Make a pledge button to allow additional pledges)

- [ ] Clicking on Increase Pledge, does not redirect back to Main Pledge Screen

- [ ] Coming back to home from pledge event does not show the pending pledges, the app needds to be restarted

- [ ] Clicking donate in the unfulfilled pledge (for now -redirect to project/ donate page)

- [ ] Commas on unfulfilled pledge numbers

- [ ] Increasing pledge, (you can actually decrease it to -500) (pledge amount should be higher than before . check validation)

File: app/server/routes/fos_js_routes.json [hunk 1]
```diff
@@ -2722,6 +2722,39 @@
 			],
 			"schemes": []
 		},
+		"eventPledgeAuthed_form": {
```
Comment (norbertschuler) [inline app/server/routes/fos_js_routes.json]: These missing API calls are already added in the development branch, but could be shown as merge conflicts due to different formatting of the file. I suggest you just take the version of the develop branch.

File: app/server/formSchemas/pledge.js [hunk 2]
```diff
@@ -45,7 +45,8 @@ export default {
       pattern: '.{1,}',
       icon: null,
       help: null,
-      propertyOrder: 4
+      propertyOrder: 4,
+      minimum: 1
```
Comment (norbertschuler) [inline app/server/formSchemas/pledge.js]: Is this change reflected on the backend too?

File: ios/Podfile [hunk 3]
```diff
@@ -3,4 +3,8 @@
 
 target 'TreecounterApp' do
   inherit! :search_paths
+  pod 'RNVectorIcons', :path => '../node_modules/react-native-vector-icons'
+
+  pod 'RNCAsyncStorage', :path => '../node_modules/@react-native-community/async-storage'
```
[the remark this answers is not in the document]
Comment (norbertschuler) [reply ios/Podfile]: I still think this should be changed. There is a documentation of manual installation of a node module for iOS at https://github.com/oblador/react-native-vector-icons#ios

File: package.json [hunk 4]
```diff
@@ -75,6 +76,7 @@
     "react-native-svg": "^6.5.3",
     "react-native-svg-uri": "^1.2.3",
     "react-native-tab-view": "^1.0.2",
+    "react-native-vector-icons": "^6.6.0",
```
[the remark this answers is not in the document]
Comment (norbertschuler) [reply package.json]: I suggest to delete all the unused fonts (e.g. all except `FontAwesome*`) like it was done in https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1460 (i.d. remove from PR and also remove from iOS project file and Info.plist et al).

File: web/config/webpack.dev.config.js [hunk 5]
```diff
@@ -5,27 +5,27 @@ const path = require('path');
...
+    //   path.join(__dirname, '../widgets/DonateTrees/widget.js')
+    // ],
+    // progressbarwidget: [
+    //   'babel-polyfill',
+    //   path.join(__dirname, '../widgets/progressbar/widget.js')
+    // ],
+    // ndviwidget: [
+    //   'babel-polyfill',
+    //   path.join(__dirname, '../widgets/NDVI/widget.js')
+    // ]
```
Comment (norbertschuler) [inline web/config/webpack.dev.config.js]: This shouldn't be commented for the release version!

File: package.json [hunk 6]
```diff
@@ -35,6 +35,7 @@
     "classnames": "^2.2.5",
     "css-loader": "^0.28.9",
     "esri-loader": "^2.2.0",
+    "formik": "^1.5.8",
```
Comment (norbertschuler) [inline package.json]: This PR introduces `formik` just next to `tcomb-form` making this project a mix of different form libraries. I do not like that as it introduces more complexity for new developers again.

File: package.json [hunk 7]
```diff
@@ -105,7 +108,8 @@
     "tcomb-json-schema": "^0.3.2",
     "tipsi-stripe": "^5.6.0",
     "uuidv4": "^2.0.0",
-    "why-did-you-update": "1.0.6"
+    "why-did-you-update": "1.0.6",
+    "yup": "^0.27.0"
```
Comment (norbertschuler) [inline package.json]: I assume this belong to `formik` adding another 38k of JavaScript to our project :-)

File: package.json [hunk 8]
```diff
@@ -78,11 +79,13 @@
     "react-native-svg": "^6.5.3",
     "react-native-svg-uri": "^1.2.3",
     "react-native-tab-view": "^1.0.2",
+    "react-native-vector-icons": "^6.6.0",
     "react-native-youtube": "^1.1.0",
     "react-navigation": "^2.11.2",
     "react-navigation-redux-helpers": "^1.1.2",
     "react-notifications": "^1.4.3",
     "react-redux": "^7.1.0",
+    "react-reveal": "^1.2.2",
```
Comment (norbertschuler) [inline package.json]: This does not seem to be used. Eslint for sure will give an error for unused imports when we try to merge this PR into develop or via verse. We should remove unused stuff.

File: app/server/routes/fos_js_routes.json (hunk 1 above)
Comment (harshvitra) [reply app/server/routes/fos_js_routes.json]: I took the latest version

File: web/config/webpack.dev.config.js (hunk 5 above)
Comment (harshvitra) [reply web/config/webpack.dev.config.js]: Uncommented

File: package.json (hunk 6 above)
Comment (harshvitra) [reply package.json]: We are slowly migrating to Formik as Tcomb-form works separately for web and separate library for app. 

Formik can be used for both and we already are struggling with the understanding of how the schema files are generated so would not want to rely on them

File: android/app/build.gradle [hunk 9]
```diff
@@ -151,6 +151,8 @@ android {
 }
 
 dependencies {
+    implementation project(':@react-native-community_async-storage')
```
Comment (norbertschuler) [inline android/app/build.gradle]: This is a duplicate as it is already added used a few lines below. The Android project would probably fail.

File: android/settings.gradle [hunk 10]
```diff
@@ -1,4 +1,8 @@
 rootProject.name = 'TreecounterApp'
+include ':@react-native-community_async-storage'
+project(':@react-native-community_async-storage').projectDir = new File(rootProject.projectDir, '../node_modules/@react-native-community/async-storage/android')
```
Comment (norbertschuler) [inline android/settings.gradle]: This is a duplicate as it is already added used a few lines below. The Android project would probably fail.

File: app/actions/pledgeEventsAction.js [hunk 11]
```diff
@@ -1,20 +1,25 @@
...
-    //     setPledgeEvents(
-    //       val.data.sort(function(a, b) {
-    //         return a.position < b.position;
-    //       })
-    //     )
-    //   )
-    // );
+    getRequest('public_pledgeEvents_get').then(val =>
+      dispatch(setPledgeEvents(val.data))
+    );
```
Comment (norbertschuler) [inline app/actions/pledgeEventsAction.js]: I recently added `.catch(error => console.log(error));` to all promises (at least the ones with possible network failures) - please do as well.

File: app/actions/pledgeEventsAction.js [hunk 12]
```diff
@@ -1,20 +1,25 @@
...
+export function fetchPublicPledgesAction(pledgeTokens) {
+  console.log('Function Called');
+  return dispatch => {
+    getRequest('eventPledgesByTokenPublic_get', {
+      version: 'v1.3',
+      pledgeTokens: pledgeTokens
+    }).then(res => {
+      dispatch(mergeEntities(normalize(res.data, [eventPledgeSchema])));
+      console.log('Got event by token');
+    });
```
Comment (norbertschuler) [inline app/actions/pledgeEventsAction.js]: I recently added `.catch(error => console.log(error));` to all promises (at least the ones with possible network failures) - please do as well.

File: app/components/UserContributions/ContributionCard.js [hunk 13]
```diff
@@ -315,7 +315,9 @@ export default class ContributionCard extends React.Component {
             {treeCountLine ? <TextSpan strong>{treeCountLine}</TextSpan> : null}
             {plantProjectLine ? <TextSpan>{plantProjectLine}</TextSpan> : null}
             {plantActionLine ? <TextSpan>{plantActionLine}</TextSpan> : null}
-            {dedicateActionLine ? <TextSpan>{dedicateActionLine}</TextSpan> : null}
+            {dedicateActionLine ? (
+              <TextSpan>{dedicateActionLine}</TextSpan>
+            ) : null}
```
Comment (norbertschuler) [inline app/components/UserContributions/ContributionCard.js]: I wonder why you reverted my change here? I changed this to `{dedicateActionLine ? <TextSpan>{dedicateActionLine}</TextSpan> : null}` - is there anything wrong with it or is this a merge error?

Comment (harshvitra) [reply app/components/UserContributions/ContributionCard.js]: I'm sure that I didn't touch this file

File: app/actions/pledgeAction.js [hunk 14]
```diff
@@ -33,24 +48,95 @@ export function fetchPledgesAction(eventSlug) {
...
+  return () => {
+    loggedIn
+      ? putAuthenticatedRequest('eventPledgeAuthed_put', data, params).then(
+          res => {
+            return res.data;
+          }
+        )
+      : putRequest('eventPledge_put', data, params).then(res => {
+          return res.data;
+        });
```
[the remark this answers is not in the document]
Comment (harshvitra) [reply app/actions/pledgeAction.js]: Done

File: app/server/formSchemas/pledge.js (hunk 2 above)
Comment (norbertschuler) [reply app/server/formSchemas/pledge.js]: @jmiridis how is this file created in the backend, is it?

File: app/actions/pledgeEventsAction.js (hunk 11 above)
Comment (harshvitra) [reply app/actions/pledgeEventsAction.js]: Done

File: app/locales/en/pledgelabels.json [hunk 15]
```diff
@@ -16,6 +16,16 @@
...
+  "treesPledgedAllPledges": "{{treeCount}} trees pledged",
+  "unfulfilledPledgesTitle": "Unfulfilled Pledges",
+  "anonymous": "Anonymous",
+  "eventDescriptionAbout": "About",
+  "increaseMyPledge": "+ INCREASE MY PLEDGE",
+  "increasePledge": "Increase Pledge",
+  "allPledges": "All Pledges",
+  "treesPledgedOn": "List of Trees Pledged on",
+  "donateXTrees": "Donate {{treeCount}} Trees",
+  "increasePledgeMessage": "To increase your pledge, please enter an amount higher than {{treeCount}} trees. A tree costs {{treeCost}} {{currency}} and are planted in {{projectName}}"
```
[the remark this answers is not in the document]
Comment (harshvitra) [reply app/locales/en/pledgelabels.json]: I dont have login access

File: app/components/UserContributions/ContributionCard.js (hunk 13 above)
Comment (norbertschuler) [reply app/components/UserContributions/ContributionCard.js]: ok, sorry, that was automatically done

File: app/locales/en/pledgelabels.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (norbertschuler) [reply app/locales/en/pledgelabels.json]: Well, you don't have to do the German translation. This comment I just did to make us aware we have to wait for this additional step before we can release the web client with this PR.

File: package.json [hunk 16]
```diff
@@ -36,11 +36,13 @@
     "css-loader": "^0.28.9",
     "date-fns": "^2.4.1",
     "esri-loader": "^2.2.0",
+    "formik": "^1.5.8",
     "i18next": "^11.3.1",
     "intl": "1.2.5",
     "jwt-decode": "^2.2.0",
     "liform-react": "^0.9.0",
     "lodash": "^4.17.10",
+    "moment": "^2.24.0",
```
Comment (norbertschuler) [inline package.json]: This has been replaced with `date-fns`. I will change that again.

File: app/server/formSchemas/pledge.js (hunk 2 above)
Comment (norbertschuler) [reply app/server/formSchemas/pledge.js]: Btw. what exactly does this line do?

**Comment by norbertschuler** [comment]
I now tested the functionality and found two issues (related):
- If I make more than one donation to a single pledge event and then use the button to increase my donation, only the first donation always get increased.
- If I make a donation and navigate to another pledge events, the button to donate or increase my pledge are still shown and still related to the first visited pledge event. After visiting a third pledge event, I get again the button to make a new pledge.

**Comment by norbertschuler** [comment]
I created a clean branch without the `react-native-vector-icons`  module at https://github.com/Plant-for-the-Planet-org/treecounter-app/tree/feature/unfulfilled_pledges_2 - feel free to use it.

File: ios/Podfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (norbertschuler) [reply ios/Podfile]: resolves with https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1555

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (norbertschuler) [reply package.json]: resolved with https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1555

**Pull request closed without merging**